### ISSUE TITLE
no examples on google as how to actually instantiate a 2.0. Entity

### ISSUE THREAD

**Issue by johnda98** [MorphiaOrg/morphia#1498]
presume these examples are missing their create method ?  author name   and for bookname

ie Author(String name) (
this.name = name;
}

AND... how do we set the relationship to Book.. we choose EITHER ? List, Set or Map ??

Can you PLEASE add a add new books to Author example

```java
private class Author {
    @Id
    private ObjectId id;
    private String name;
    private MorphiaReference<List<Book>> list;
    private MorphiaReference<Set<Book>> set;
    private MorphiaReference<Map<String, Book>> map;

    public Author() { }

    public List<Book> getList() {
        return list.get();
    }

    public void setList(final List<Book> list) {
        this.list = MorphiaReference.wrap(list);
    }

    public Set<Book> getSet() {
        return set.get();
    }

    public void setSet(final Set<Book> set) {
        this.set = MorphiaReference.wrap(set);
    }

    public Map<String, Book> getMap() {
        return map.get();
    }

    public void setMap(final Map<String, Book> map) {
        this.map = MorphiaReference.wrap(map);
    }
}

private class Book {
    @Id
    private ObjectId id;
    private String name;
    private MorphiaReference<Author> author;

    public Book() { }

    public Author getAuthor() {
        return author.get();
    }

    public void setAuthor(final Author author) {
        this.author = MorphiaReference.wrap(author);
    }
}
```

**Comment by evanchooly** [comment]
Whether you choose List, Set, or Map depends entirely on how you want to model those relationships.  Usually such relationships are done with Lists.

**Comment by johnda98** [comment]
Thanks - that helps.. I think we'll back pedal to 1.4 from 2.0... and thanks also to Justin.    for some reason we has trouble with .wrap just trying to get an instance from the list..   probably best if we stick with 1.4 until we learn more here

**Comment by evanchooly** [comment]
`wrap()` just takes the `List<Book>` reference and creates the appropriate subclass of `MorphiaReference`.  That subclass is irrelevant to you as a coder and just has the logic needed to process the `List<Book>`.  `wrap()` is overload so that you needn't care necessarily what reference type you're passing in or what the underlying `MorphiaReference` implementation is using.  You can also, of course, just use the traditional `@Reference` annotation like you would in 1.4.

Though 1.4 is rather old and you'd be missing out on a lot of bug fixes to go back that far.

**Comment by johnda98** [comment]
thanks Justin.. that helps..  we got scared off 2.0 due to lack of
examples..   BUT it seems 1.4's morphia.createDatastore uses
com.mongodb.MongoClient  and the ONLY driver that Atlas mongo will use,
uses the dependency   com.mongodb.client.MongoClient ..     and we wanted
to use Atlas cluster  not a local mongo

so I guess I wil have to go back to morphia 2.0 and do trial and error..
any github examples out there that uses 2.0 extensively on a large schema
?  google search didnt pull hardly any up

UNLESS you know of a driver that Atlas and its connection  'needs'  will
also enable   morphia.createDatastore to be happy  ie the acceptable
com.mongodb.client.MongoClient



thanks man  u guys rock !

*John D Allen.*


*T +1 417 849 0705 ** CST  | john.allen@leveridgesystems.com
<john.allen@leveridgesystems.com>  | https://www.leveridgesystems.com
<https://www.leveridgesystems.com> | linkedin.com/in/johndallen1
<http://linkedin.com/in/johndallen1> | skype john.allen_67
https://www.yourhashgraphtoken.com <https://www.yourhashgraphtoken.com>*


 Leveridge Systems INC. Reg Missouri Charter #01082460.  The information
contained in this email and any attached files is confidential and intended
solely for the addressee(s). The email may be legally privileged or
prohibited from disclosure and unauthorized use. If you are not the
named addressee you may not use, copy, or disclose this information to any
other person. If you received this message in error please notify the
sender immediately and permanently delete it from your system.

**Comment by johnda98** [comment]
any links to LOTS of 2.0 java code examples ?  as polyglot its 10 times as
fast to get cookin than reading docs.. we just extrapolate from the examples

thanks if any links.

yes.. Atlas connection uses a driver that I think doesnt support
morphia.createDatastore for 1.4 anyways



*John D Allen.*


*T +1 417 849 0705 ** CST  | john.allen@leveridgesystems.com
<john.allen@leveridgesystems.com>  | https://www.leveridgesystems.com
<https://www.leveridgesystems.com> | linkedin.com/in/johndallen1
<http://linkedin.com/in/johndallen1> | skype john.allen_67
https://www.yourhashgraphtoken.com <https://www.yourhashgraphtoken.com>*


 Leveridge Systems INC. Reg Missouri Charter #01082460.  The information
contained in this email and any attached files is confidential and intended
solely for the addressee(s). The email may be legally privileged or
prohibited from disclosure and unauthorized use. If you are not the
named addressee you may not use, copy, or disclose this information to any
other person. If you received this message in error please notify the
sender immediately and permanently delete it from your system.

**Comment by evanchooly** [comment]
You might consider 1.6 as its API is closer to what you're used to with 1.4 but was released, in part, to bring support to the 4.0 driver series to the 1.x line of Morphia.

As for examples, as always docs lag behind intent.  My hope is to create a cookbook type repository of examples (taken from, say, stackoverflow questions or examples in the main MongoDB docs) but I simply haven't had that time yet.  In general, I've tried to make the 2.0 API mirror the server side components (especially with filters and aggregation expressions) so there'd be less mental translation when moving back and forth.

**Comment by johnda98** [comment]
null pointer exception.. BONE simple code...    I dont think 1.6 will work
with Atlas as if its like 1.4 and DOES not use below.. which Atlas uses ..
ie the  .client.    1.6 uses com.mongodb.Mongoclient and
com.mongodb.Clients.


import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;


anyways..  back to 2.0.1  VERY cool !.. so I would implement a Bone simple
Author Book.. from the Morphia site..    BUT it could not add the
relationship between books of the author... got a null point on the .add
call

getting bald here.. hair pullin/falling out fast

public void runquery(ActionEvent actionEvent) throws IOException {
  //  MongoConnection conn = MongoConnection.getInstance();
    MongoConnectiontest conn = MongoConnectiontest.getInstance();

    conn.init();

    Datastore datastore = conn.getDatastore();

    // test MorphiaReference on Author and Book 1:M  to test my morpha
class loader etc

    // create author and 2 books

    final  Author author1 = new Author("Jane");

    datastore.save(author1);

    final  Author author2 = new Author("Bob");

    datastore.save(author2);

    // create 2 books by Jane

    final Book book1 = new Book("Jane's dog");

    datastore.save(book1);

    final Book book2 = new Book("Jane's cat");

    datastore.save(book2);



    // save the relation

    author1.getList().add(book1);
    author1.getList().add(book2);

    // re-save author 1 and both books again ?? guess so .. no
documentation on that

    datastore.save(author1);

    datastore.save(book1);
    datastore.save(book2);
    // ok query Jane's book .. jeeeezus  get her first !

   // Query<Author> getauth = datastore.find(Author.class);

  // System.out.println("found Jane ?" + getauth.first().toString());




entities....


import java.util.Set;

@Entity("authors")
public class Author {
    @Id
    private ObjectId id;
    private String name;
    private MorphiaReference<List<Book>> list;
    private MorphiaReference<Set<Book>> set;
    private MorphiaReference<Map<String, Book>> map;

    public Author() { }

    public Author(String name) {
        this.name = name;
    }

    public List<Book> getList() {
        return list.get();
    }

    public void setList(final List<Book> list) {
        this.list = MorphiaReference.wrap(list);
    }

    public Set<Book> getSet() {
        return set.get();
    }

    public void setSet(final Set<Book> set) {
        this.set = MorphiaReference.wrap(set);
    }

    public Map<String, Book> getMap() {
        return map.get();
    }

    public void setMap(final Map<String, Book> map) {
        this.map = MorphiaReference.wrap(map);
    }
}




package org.crob;

import dev.morphia.annotations.Entity;
import dev.morphia.annotations.Id;
import dev.morphia.mapping.experimental.MorphiaReference;
import org.bson.types.ObjectId;

@Entity("books")
public class Book {
    @Id
    private ObjectId id;
    private String name;
    private MorphiaReference<Author> author;

    public Book() { }

    public Book(String name) {
        this.name = name;
    }

    public Author getAuthor() {
        return author.get();
    }

    public void setAuthor(final Author author) {
        this.author = MorphiaReference.wrap(author);
    }
}





Starting DataStore on DB: test
Exception in thread "JavaFX Application Thread" java.lang.RuntimeException:
java.lang.reflect.InvocationTargetException
at
javafx.fxml/javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1787)
at
javafx.fxml/javafx.fxml.FXMLLoader$ControllerMethodEventHandler.handle(FXMLLoader.java:1670)
at
javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
at
javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
at
javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
at
javafx.base/com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
at
javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
at
javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
at
javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
at
javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
at
javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
at
javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
at
javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
at javafx.base/javafx.event.Event.fireEvent(Event.java:198)
at javafx.graphics/javafx.scene.Node.fireEvent(Node.java:8879)
at javafx.controls/javafx.scene.control.Button.fire(Button.java:200)
at
javafx.controls/com.sun.javafx.scene.control.behavior.ButtonBehavior.mouseReleased(ButtonBehavior.java:206)
at
javafx.controls/com.sun.javafx.scene.control.inputmap.InputMap.handle(InputMap.java:274)
at
javafx.base/com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
at
javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
at
javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
at
javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
at
javafx.base/com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
at
javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
at
javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
at
javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
at
javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
at
javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
at
javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
at
javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
at javafx.base/javafx.event.Event.fireEvent(Event.java:198)
at javafx.graphics/javafx.scene.Scene$MouseHandler.process(Scene.java:3851)
at
javafx.graphics/javafx.scene.Scene$MouseHandler.access$1200(Scene.java:3579)
at javafx.graphics/javafx.scene.Scene.processMouseEvent(Scene.java:1849)
at
javafx.graphics/javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2588)
at
javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:397)
at
javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:295)
at java.base/java.security.AccessController.doPrivileged(Native Method)
at
javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$2(GlassViewEventHandler.java:434)
at
javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:390)
at
javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:433)
at javafx.graphics/com.sun.glass.ui.View.handleMouseEvent(View.java:556)
at javafx.graphics/com.sun.glass.ui.View.notifyMouse(View.java:942)
at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication._runLoop(Native
Method)
at
javafx.graphics/com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$11(GtkApplication.java:277)
at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.reflect.InvocationTargetException
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native
Method)
at
java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at
java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:566)
at com.sun.javafx.reflect.Trampoline.invoke(MethodUtil.java:76)
at jdk.internal.reflect.GeneratedMethodAccessor3.invoke(Unknown Source)
at
java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:566)
at javafx.base/com.sun.javafx.reflect.MethodUtil.invoke(MethodUtil.java:273)
at javafx.fxml/com.sun.javafx.fxml.MethodHelper.invoke(MethodHelper.java:83)
at
javafx.fxml/javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1782)
... 47 more
Caused by: java.lang.NullPointerException
at org.crob/org.crob.Author.getList(Author.java:28)
at
org.crob/org.crob.SecondaryController.runquery(SecondaryController.java:60)
... 58 more







*John D Allen.*


*T +1 417 849 0705 ** CST  | john.allen@leveridgesystems.com
<john.allen@leveridgesystems.com>  | https://www.leveridgesystems.com
<https://www.leveridgesystems.com> | linkedin.com/in/johndallen1
<http://linkedin.com/in/johndallen1> | skype john.allen_67
https://www.yourhashgraphtoken.com <https://www.yourhashgraphtoken.com>*


 Leveridge Systems INC. Reg Missouri Charter #01082460.  The information
contained in this email and any attached files is confidential and intended
solely for the addressee(s). The email may be legally privileged or
prohibited from disclosure and unauthorized use. If you are not the
named addressee you may not use, copy, or disclose this information to any
other person. If you received this message in error please notify the
sender immediately and permanently delete it from your system.
